petastorm/spark/spark_dataset_converter.py

```diff
@@ -0,0 +1,310 @@
+                                         compression_codec):
+    """
+    Check whether the df is cached.
+    If so, return the existing cache file path.
+    If not, cache the df into the cache_dir in parquet format and return the
+    cache file path.
+    Use atexit to delete the cache before the python interpreter exits.
+    :param df: A :class:`DataFrame` object.
+    :param parent_cache_dir_url: A string denoting the directory for the saved
+        parquet file.
+    :param parquet_row_group_size_bytes: An int denoting the number of bytes
+        in a parquet row group.
+    :param compression_codec: Specify compression codec.
+    :return: A string denoting the path of the saved parquet file.
+    """
+    # TODO
+    #  Improve the cache list by hash table (Note we need use hash(df_plan +
+    #  row_group_size)
+    with _cache_df_meta_list_lock:
+        df_plan = _get_df_plan(df)
+        for meta in _cache_df_meta_list:
+            if meta.row_group_size == parquet_row_group_size_bytes and \
+                    meta.compression_codec == compression_codec and \
+                    meta.df_plan.sameResult(df_plan):
+                return meta.cache_dir_url
+        # do not find cached dataframe, start materializing.
+        cached_df_meta = CachedDataFrameMeta.create_cached_dataframe(
+            df, parent_cache_dir_url, parquet_row_group_size_bytes,
+            compression_codec)
+        _cache_df_meta_list.append(cached_df_meta)
+        return cached_df_meta.cache_dir_url
+
+
+def _materialize_df(df, parent_cache_dir_url, parquet_row_group_size_bytes,
+                    compression_codec):
+    uuid_str = str(uuid.uuid4())
+    save_to_dir_url = _make_sub_dir_url(parent_cache_dir_url, uuid_str)
+
+    df.write \
+        .option("compression", compression_codec) \
+        .option("parquet.block.size", parquet_row_group_size_bytes) \
+        .parquet(save_to_dir_url)
+
+    logging.info('Materialize dataframe to url %s successfully.', save_to_dir_url)
+
+    atexit.register(_delete_cache_data_atexit, save_to_dir_url)
+
+    return save_to_dir_url
+
+
+def make_spark_converter(
+        df,
+        parquet_row_group_size_bytes=DEFAULT_ROW_GROUP_SIZE_BYTES,
+        compression_codec=None):
+    """
+    Convert a spark dataframe into a :class:`SparkDatasetConverter` object.
+    It will materialize a spark dataframe to a `cache_dir_url`.
+    The returned `SparkDatasetConverter` object will hold the materialized
+    dataframe, and can be used to make one or more tensorflow datasets or
+    torch dataloaders.
+    We can explicitly delete the materialized dataframe data, see
+    `SparkDatasetConverter.delete`, and when the spark application exit,
+    it will try best effort to delete the materialized dataframe data.
+
+    :param df: The :class:`DataFrame` object to be converted.
+    :param parquet_row_group_size_bytes: An int denoting the number of bytes
+        in a parquet row group.
```

Comment:
It was not clear from the beginning of the comment that we are materialing the dataframe into a parquet store. We should mention parquet before so the meaning of this and the next argument become clear.